### DISCUSSION TITLE
Behavior of make before break

### DISCUSSION THREAD

**Discussion by amvarunn** [strongswan/strongswan#852] [Q&A]
We are trying to enable make_before_break and we have set the corresponding flag in the charon section in strongswan.conf. Please find the attached charon logs and help in understanding some question that follows.

1: Fri Jan 14 21:02:31 2022 05[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1161> reauthenticating IKE_SA ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868[1161]

**_2: Fri Jan 14 21:02:31 2022 05[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1161> initiating IKE_SA ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868[1162] to 192.168.56.14_**

3: Fri Jan 14 21:02:31 2022 13[CFG] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_AES128_XCBC/ECP_256

4: Fri Jan 14 21:02:31 2022 13[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> received cert request for "C=x, O=y, CN=Root CA"

5: Fri Jan 14 21:02:31 2022 13[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> sending cert request for "C=x, O=y, CN=Root CA"

6: Fri Jan 14 21:02:31 2022 13[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> authentication of '76:3b:48:ee:ae:da:11:f8:49:79:3c:85:e9:f5:2c:d6:ea:45:20:0a' (myself) with RSA_EMSA_PKCS1_SHA2_256 successful

7: Fri Jan 14 21:02:31 2022 13[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> sending end entity cert "C=x, ST=a, L=b, O=y, OU=Systems, CN=SVC_CA_Signed_cluster02, E=support@example.com"

8: Fri Jan 14 21:02:31 2022 13[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> establishing CHILD_SA ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868{924} reqid 1

9: Fri Jan 14 21:02:31 2022 16[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1161> received DELETE for IKE_SA ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868[1161]

10: Fri Jan 14 21:02:31 2022 16[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1161> deleting IKE_SA ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868[1161] between 192.168.56.11[76:3b:48:ee:ae:da:11:f8:49:79:3c:85:e9:f5:2c:d6:ea:45:20:0a]...192.168.56.14[5b:5c:77:b7:84:55:ad:39:7a:49:b3:34:18:c4:f0:ae:84:2b:f4:e7]

11: Fri Jan 14 21:02:31 2022 16[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1161> restarting CHILD_SA ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868

12: Fri Jan 14 21:02:31 2022 16[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1161> initiating IKE_SA ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868[1163] to 192.168.56.14

**_13: Fri Jan 14 21:02:31 2022 16[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1161> IKE_SA deleted_**

14: Fri Jan 14 21:02:31 2022 16[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> received end entity cert "C=x, ST=a, L=b, O=y, OU=Systems, CN=SVC_CA_Signed_cluster01, E=support@example.com"

15: Fri Jan 14 21:02:31 2022 16[CFG] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162>   using certificate "C=x, ST=a, L=b, O=y, OU=Systems, CN=SVC_CA_Signed_cluster01, E=support@example.com"

16: Fri Jan 14 21:02:31 2022 16[CFG] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162>   using trusted ca certificate "C=x, O=y, CN=Root CA"

17: Fri Jan 14 21:02:31 2022 16[CFG] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162>   reached self-signed root ca with a path length of 0

18: Fri Jan 14 21:02:31 2022 16[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> authentication of '5b:5c:77:b7:84:55:ad:39:7a:49:b3:34:18:c4:f0:ae:84:2b:f4:e7' with RSA_EMSA_PKCS1_SHA2_256 successful

_**19: Fri Jan 14 21:02:31 2022 16[IKE] <ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868|1162> IKE_SA ED_USR_IPR_151_0_0_2_65535_IPVX_192-168-56-11_192-168-56-14_3868[1162] established between 192.168.56.11[76:3b:48:ee:ae:da:11:f8:49:79:3c:85:e9:f5:2c:d6:ea:45:20:0a]...192.168.56.14[5b:5c:77:b7:84:55:ad:39:7a:49:b3:34:18:c4:f0:ae:84:2b:f4:e7]**_

In  Line2    IKE_SA 1162 initiate begins, 
     Line13  IKE_SA 1161 is deleted and 
     Line19  IKE_SA 1162 is established.

Question 1: Is this the expected behaviour? Or, do we expect 1162 to be established, before 1161 is deleted? 
Question2: The IKE_SA numbers (1162, 1163 etc) sometimes does not appear sequential. Is that expected?

Strongswan.conf file
[strongswanconf_ssupload.txt](https://github.com/strongswan/strongswan/files/7869566/strongswanconf_ssupload.txt)

Thank you

**Comment by tobiasbrunner** [comment]
> Is this the expected behaviour

No,  your peer deletes the old IKE_SA, it shouldn't. Check the log there.

> The IKE_SA numbers (1162, 1163 etc) sometimes does not appear sequential. Is that expected?

Possible if there are other IKE_SAs created in between.

**Comment by amvarunn** [comment]
We have reproduced this issue, and in the remote it appears establish is happening before delete.
However, please help in understanding why is there a duplicate SA (Line 139 and 143) even though we are trying make before break?
[query.txt](https://github.com/strongswan/strongswan/files/7887353/query.txt)

**Reply by tobiasbrunner** [reply]
You  have the [_duplicheck_ plugin](https://wiki.strongswan.org/projects/strongswan/wiki/Duplicheck) loaded, you don't want that. So make sure you disable it.
